plugin/forward/connect.go

```diff
@@ -91,6 +91,15 @@ func (p *Proxy) Connect(ctx context.Context, state request.Request, opts options
 		proto = state.Proto()
 	}
 
+	var err error
+	defer func() {
+		if err != nil && err != dns.ErrTruncated && err != ErrCachedClosed {
+			RequestFailureCount.WithLabelValues(p.addr).Add(1)
+		}
+	}()
+
+	RequestCount.WithLabelValues(p.addr).Add(1)
```

Comment:
What is actually the counters RequestCount and RequestFailureCount intended to count?
I believe, they should show how many **user** requests were handled/failed. In other words, how many times ServeDNS() was invoked successfully or not successfully.
The counters incremented here don't give us these numbers. Method Connect() can be called many times from the same ServeDNS() invocation because of retrying on some errors.
Perhaps, we need to update these metrics at higher level, i.e. in ServeDNS() itself.

Reply:
> What is actually the counters RequestCount and RequestFailureCount intended to count? I believe, they should show how many user requests were handled/failed.

I agree user requests should be measured at the highest level, ie. `core.ServeDNS`.

I don't think plugin metrics (at least as they are implemented now) were ever intended to measure user requests but rather provide plugin-level insight into state of the DNS service.

E.g.: https://github.com/coredns/coredns/blob/master/plugin/proxy/README.md
> coredns_proxy_request_count_total{server, proto, proto_proxy, family, to} - query count per upstream.

Both levels of metrics make sense, i want user request metrics to determine quality of service as seen by users. But i also want plugin-level metrics for troubleshooting and early detection of issues that might other wise be masked because of retrying (e.g. network congestion, dropped upstream requests).

Reply:
Yeah let me improve the documentation of these metrics to make clear that they not necessarily correspond to rate of user requests.